src/source/publishing/influx_publisher.cr

```diff
@@ -36,48 +55,124 @@ module PlaceOS::Source
       end
     end
 
+    NO_POINTS = [] of Flux::Point
+
     # Generate an InfluxDB Point from an mqtt key + payload
     #
-    def self.transform(message : Publisher::Message, timestamp : Time = Publisher.timestamp) : Flux::Point?
+    # ameba:disable Metrics/CyclomaticComplexity
+    def self.transform(message : Publisher::Message, timestamp : Time = Publisher.timestamp) : Array(Flux::Point)
       data = message.data
       # Only Module status events are persisted
-      return unless data.is_a? Mappings::Status
+      return NO_POINTS unless data.is_a? Mappings::Status
 
       payload = message.payload.try &.gsub(Regex.union(DEFAULT_FILTERS)) do |match_string, _|
         hmac_sha256(match_string)
       end
 
-      tags = Hash(Symbol, String?){
-        :org      => data.zone_mapping["org"],
-        :building => data.zone_mapping["building"],
+      # Namespace tags and fields to reduce likely hood that they clash with status names
+      tags = Hash(String, String?){
+        "pos_org"      => data.zone_mapping["org"],
+        "pos_building" => data.zone_mapping["building"],
       }.compact
+      fields = ::Flux::Point::FieldSet.new
+
+      # https://docs.influxdata.com/influxdb/v2.0/reference/flux/language/lexical-elements/#identifiers
+      key = data.status.gsub(/\W/, '_')
+      fields["pos_key"] = key
+
+      # Influx doesn't support nil values
+      if !payload
+        Log.info { {message: "Influx doesn't support nil values", status: key} }
+        return NO_POINTS
+      end
+
+      begin
+        raw = Value.from_json(payload)
+
+        case raw
+        in Flux::Point::FieldType
+          fields[key] = raw
+        in Hash(String, Flux::Point::FieldType?)
+          compacted = raw.compact
+          return NO_POINTS if compacted.empty?
+
+          compacted.each do |sub_key, value|
+            sub_key = sub_key.gsub(/\W/, '_')
+            fields[sub_key] = value
+          end
+        in CustomMetrics
+          # Add the tags and fields going to all points
```

Comment:
it might be easier to follow if this block is extracted into a method, as this method is slightly unwieldy. 
mainly in regards to a single large case returning, while the others are a mix of early returns and continued execution